raphtory/src/python/graph/history.rs

```diff
@@ -0,0 +1,793 @@
+    pub fn merge(&self, other: &Self) -> Self {
+        Self {
+            history: History::new(Arc::new(MergedHistory::new(
+                self.history.0.clone(),
+                other.history.0.clone(),
+            ))),
+        }
+    }
+
+    // Clones the Arcs, we end up with Arc<Arc<dyn InternalHistoryOps>>, 1 level of indirection. Cloning the underlying InternalHistoryOps objects introduces lifetime issues.
+    pub fn reverse(&self) -> Self {
+        PyHistory {
+            history: History::new(Arc::new(ReversedHistoryOps::new(self.history.0.clone()))),
+        }
+    }
+
+    /// Access history events as i64 timestamps
+    #[getter]
+    pub fn t(&self) -> PyHistoryTimestamp {
+        PyHistoryTimestamp {
+            history_t: HistoryTimestamp::new(self.history.0.clone()), // clone the Arc, not the underlying object
+        }
+    }
+
+    /// Access history events as DateTime items
+    #[getter]
+    pub fn dt(&self) -> PyHistoryDateTime {
+        PyHistoryDateTime {
+            history_dt: HistoryDateTime::new(self.history.0.clone()), // clone the Arc, not the underlying object
+        }
+    }
+
+    /// Access secondary index (unique index) of history events
+    #[getter]
+    pub fn secondary_index(&self) -> PyHistorySecondaryIndex {
+        PyHistorySecondaryIndex {
+            history_s: HistorySecondaryIndex::new(self.history.0.clone()), // clone the Arc, not the underlying object
+        }
+    }
+
+    #[getter]
+    pub fn intervals(&self) -> PyIntervals {
+        PyIntervals {
+            intervals: Intervals::new(self.history.0.clone()), // clone the Arc, not the underlying object
+        }
+    }
+
+    /// Get the earliest time in the history
+    pub fn earliest_time(&self) -> Option<TimeIndexEntry> {
+        self.history.earliest_time()
+    }
+
+    /// Get the latest time in the history
+    pub fn latest_time(&self) -> Option<TimeIndexEntry> {
+        self.history.latest_time()
+    }
+
+    /// Collect all time events
+    pub fn collect(&self) -> Vec<TimeIndexEntry> {
+        self.history.collect()
+    }
+
+    /// Collect all time events in reverse
+    pub fn collect_rev(&self) -> Vec<TimeIndexEntry> {
+        self.history.collect_rev()
+    }
+
+    /// Iterate over all time events
+    pub fn __iter__(&self) -> PyBorrowingIterator {
+        py_borrowing_iter!(
+            self.history.clone(),
+            History<'static, Arc<dyn InternalHistoryOps>>,
+            |history| history.iter()
+        )
+    }
+
+    /// Iterate over all time events in reverse
+    pub fn __reversed__(&self) -> PyBorrowingIterator {
+        py_borrowing_iter!(
+            self.history.clone(),
+            History<'static, Arc<dyn InternalHistoryOps>>,
+            |history| history.iter_rev()
+        )
+    }
+
+    pub fn __repr__(&self) -> String {
+        self.history.repr()
+    }
+
+    fn __hash__(&self) -> u64 {
```

Comment:
probably shouldn't implement this per python conventions as the hash of the object can change!